overlord/snapstate/helpers.go

```diff
@@ -0,0 +1,132 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"errors"
+
+	"github.com/snapcore/snapd/overlord/restart"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+// Observed link order between essential snaps in doUpdate (snapstate.go)
+// base waits for snapd
+// gadget waits for base/os + snapd
+// kernel waits for base/os + snapd + gadget if present
+var essentialSnapsRestartOrder = []snap.Type{
+	snap.TypeOS,
+	// The base will only require restart if it's the base of the model.
+	snap.TypeBase,
+	snap.TypeGadget,
+	// Kernel must wait for gadget because the gadget may define
+	// new "$kernel:refs".
+	snap.TypeKernel,
+}
+
+func maybeSnapSetup(ts *state.TaskSet) *SnapSetup {
+	for _, t := range ts.Tasks() {
+		snapsup, err := TaskSnapSetup(t)
+		if err == nil {
+			return snapsup
+		}
+	}
+	return nil
+}
+
+func mappedTaskSets(tss []*state.TaskSet, bootBase string) (map[snap.Type]*state.TaskSet, error) {
+	avail := make(map[snap.Type]*state.TaskSet)
+	for _, ts := range tss {
+		snapsup := maybeSnapSetup(ts)
+		if snapsup == nil {
+			continue
+		}
+
+		switch snapsup.Type {
+		case snap.TypeBase:
+			if snapsup.SnapName() == bootBase {
+				avail[snapsup.Type] = ts
+			}
+		case snap.TypeOS, snap.TypeGadget, snap.TypeKernel:
+			avail[snapsup.Type] = ts
+		}
+	}
+	return avail, nil
+}
+
+func findUnlinkTask(ts *state.TaskSet) *state.Task {
+	for _, t := range ts.Tasks() {
+		switch t.Kind() {
+		case "unlink-snap", "unlink-current-snap":
+			return t
+		}
+	}
+	return nil
+}
+
+// SetDefaultRestartBoundaries marks "link-snap" (Do) and "unlink-snap"/"unlink-current-snap" (Undo) as restart boundaries
+// to maintain the old restart logic. This means that a restart must be performed after each of those tasks
+// in order for the change to continue.
+func SetDefaultRestartBoundaries(ts *state.TaskSet) {
```

Comment:
this seems to be used directly only in tests? could we expose only the more general one?